lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp

```diff
@@ -0,0 +1,257 @@
+
+  if (!swift_ubp.Update() || !swift_ubp.IsValid())
+    return false;
+
+  size_t count = swift_ubp.GetCount();
+  addr_t addr = swift_ubp.GetStartAddress();
+
+  stream.Printf("%zu %s (0x%" PRIx64 ")", count,
+                (count == 1) ? "value" : "values", addr);
+
+  return true;
+}
+
+namespace lldb_private {
+namespace formatters {
+namespace swift {
+class UnsafeBufferPointerSyntheticFrontEnd : public SyntheticChildrenFrontEnd {
+public:
+  UnsafeBufferPointerSyntheticFrontEnd(lldb::ValueObjectSP valobj_sp);
+
+  virtual size_t CalculateNumChildren();
+
+  virtual lldb::ValueObjectSP GetChildAtIndex(size_t idx);
+
+  virtual bool Update();
+
+  virtual bool MightHaveChildren();
+
+  virtual size_t GetIndexOfChildWithName(ConstString name);
+
+  virtual ~UnsafeBufferPointerSyntheticFrontEnd() = default;
+
+private:
+  ExecutionContextRef m_exe_ctx_ref;
+  uint8_t m_ptr_size;
+  lldb::ByteOrder m_order;
+
+  SwiftUnsafeBufferPointer m_unsafe_ptr;
+
+  std::vector<std::pair<addr_t, ValueObjectSP>> m_children;
+};
+} // namespace swift
+} // namespace formatters
+} // namespace lldb_private
+
+lldb_private::formatters::swift::UnsafeBufferPointerSyntheticFrontEnd::
+    UnsafeBufferPointerSyntheticFrontEnd(lldb::ValueObjectSP valobj_sp)
+    : SyntheticChildrenFrontEnd(*valobj_sp.get()),
+      m_unsafe_ptr(*valobj_sp.get()) {
+
+  ProcessSP process_sp = valobj_sp->GetExecutionContextRef().GetProcessSP();
+  if (!process_sp)
+    return;
+
+  if (valobj_sp)
+    Update();
+}
+
+size_t lldb_private::formatters::swift::UnsafeBufferPointerSyntheticFrontEnd::
+    CalculateNumChildren() {
+  return m_unsafe_ptr.GetCount();
+}
+
+lldb::ValueObjectSP lldb_private::formatters::swift::
+    UnsafeBufferPointerSyntheticFrontEnd::GetChildAtIndex(size_t idx) {
+  if (!m_unsafe_ptr.IsValid() && idx > m_unsafe_ptr.GetCount())
+    return ValueObjectSP();
+
+  addr_t m_start_addr = m_unsafe_ptr.GetStartAddress();
+  const uint32_t num_children = CalculateNumChildren();
+
+  if (idx >= num_children)
+    return lldb::ValueObjectSP();
+
+  if (m_children.empty()) {
+    ProcessSP process_sp = m_exe_ctx_ref.GetProcessSP();
+    if (!process_sp)
+      return lldb::ValueObjectSP();
+
+    Status error;
+    addr_t value_at_idx = 0;
+
+    uint32_t tries = 0;
+    uint32_t test_idx = 0;
+
+    // Iterate over inferior memory, reading key/value pointers by shifting each
+    // cursor by test_index * m_ptr_size. Returns an empty ValueObject if a read
+    // fails, otherwise, continue until the number of tries matches the number
+    // of childen.
+    while (tries < num_children) {
+      value_at_idx = m_start_addr + (test_idx * m_ptr_size);
+      value_at_idx = process_sp->ReadPointerFromMemory(value_at_idx, error);
+      if (error.Fail())
+        return lldb::ValueObjectSP();
+
+      test_idx++;
+
+      if (!value_at_idx)
+        continue;
+      tries++;
+
+      m_children.push_back(std::make_pair(value_at_idx, nullptr));
+    }
+  }
```

Comment:
Can this be a `for` loop? Why is a null pointer just skipped? It would create am inconsistency between the indices in memory and in the values we display which would be a huge issue.

But more generally this seems wrong? It's saying that all the values in a buffer are pointer-sized, which I'm pretty sure is wrong. The UnsafeBuffer elements are like Array elements. They are the size of their underlying type I would guess.